### PR TITLE
Deflake failures

### PR DESCRIPTION
**Pull request by BNieuwenhuizen**
This refactors the runner and then uses it to run failures twice so we can deflake.

An option is added to make flakes not failures.

### CODE REVIEW

**Comment by BNieuwenhuizen** [comment]
@anholt @robclark

**Comment by robclark** [comment]
not directly related to this MR, but some thoughts on what would make dealing w/ CI output nicer:

I tried a bit to colorize the deqp-runner.sh output to make it easier to find relevant parts, but that seems to be a dead-end.  Then I noticed that deqp already has a testlog-to-junit.  So if we could generate a unified .qpa from all the tests it would be pretty easy to get that into junit which gitlab understands.  The downsides:
1. doesn't seem to contain much detail other than that a test failed.. I'd really like some way to get back to the testlog-to-xml output for failed tests to see result images, etc
2. I guess that wouldn't understand "Flake" status.

The other option might be to generate a unified "csv" style output (which cts_runner already does), including 'Flake' status and generate the junit xml ourselves (presumably outside of the cts_runner).. which doesn't look too hard.  That would give some flexibility to (hopefully) link to testlog-to-xml output in the artifacts?

I'm not 100% sure how gitlab handles fail->fail status of a given test, but gitlab seems to imply that it does.. which might be a nice way to handle not-yet-passing tests.

File: src/cts-runner.cc [hunk 1]
```diff
@@ -471,6 +474,39 @@ void update_results(Context &ctx, std::map<std::string, enum status> results) {
   ctx.finished_cases += results.size();
 }
 
+void test_for_flakes(const Context &ctx, unsigned thread_id,
+                     const std::set<std::string>& tests,
+                     std::map<std::string, enum status> *results) {
```
Comment (robclark) [inline src/cts-runner.cc:479]: did you forget a hunk?  I don't see where this is called..

**Comment by BNieuwenhuizen** [comment]
Merged in gitlab.

**Pull request closed without merging**